src/nr_openai_observability/stream_patcher.py

```diff
@@ -0,0 +1,117 @@
+
+
+async def wrap_async_stream_generator(stream_gen):
+    async for chunk in stream_gen:
+        print("chunk----------------------------")
+        yield chunk
+
+
+def patcher_create_chat_completion_stream(original_fn, *args, **kwargs):
+    def wrap_stream_generator(stream_gen):
+        role, time_delta, content = None, None, ""
+        span = monitor.create_span()
+        try:
+            timestamp = time.time()
+            for chunk in stream_gen:
+                content += chunk.choices[0].delta.get("content", "")
+                if hasattr(chunk.choices[0].delta, "role"):
+                    role = chunk.choices[0].delta.role
+                yield chunk
+            time_delta = time.time() - timestamp
+        except Exception as ex:
+            handle_stream_completed(
+                None,
+                kwargs,
+                ex,
+                time_delta,
+                {"role": role, "content": content},
+            )
+            raise ex
+        finally:
+            span.finish()
+
+        handle_stream_completed(
+            chunk, kwargs, None, time_delta, {"role": role, "content": content}
+        )
+
+    try:
+        result = original_fn(*args, **kwargs)
+
+    except Exception as ex:
+        raise ex
+
+    wrapped_result = wrap_stream_generator(result)
+
+    return wrapped_result
+
+
+async def patcher_create_chat_completion_stream_async(original_fn, *args, **kwargs):
+    async def wrap_stream_generator(stream_gen):
+        role, time_delta, content = None, None, ""
+        span = monitor.create_span()
+        try:
+            timestamp = time.time()
+            async for chunk in await stream_gen:
+                content += chunk.choices[0].delta.get("content", "")
+                if hasattr(chunk.choices[0].delta, "role"):
+                    role = chunk.choices[0].delta.role
+                yield chunk
+            time_delta = time.time() - timestamp
+        except Exception as ex:
+            handle_stream_completed(
+                None,
+                kwargs,
+                ex,
+                time_delta,
+                {"role": role, "content": content},
+            )
+            raise ex
+        finally:
+            span.finish()
+
+        handle_stream_completed(
+            chunk, kwargs, None, time_delta, {"role": role, "content": content}
+        )
+
+    try:
+        result = original_fn(*args, **kwargs)
+
+    except Exception as ex:
+        raise ex
+
+    wrapped_result = wrap_stream_generator(result)
+
+    return wrapped_result
+
+
+@handle_errors
+def handle_stream_completed(last_chunk, request, error, response_time, message):
+    # encoding = encoding_for_model(last_chunk.get("model") or last_chunk.get("engine"))
+    # num_of_tokens = len(encoding.encode(message.get("content", "")))
+    # print(num_of_tokens)
+    events = None
+    if error:
+        events = build_completion_error_events(request, error)
```

Comment:
We may want to add a key-value pair in the error event to indicate if the error was from a streamed message.

Reply:
done